### ISSUE TITLE
componentWillUnMount never happens with StackNavigations nested inside DrawerNavigation

### ISSUE THREAD

**Issue by SamMatthewsIsACommonName** [react-navigation/react-navigation#1056]
Hi,
I've seen this issue before here https://github.com/react-community/react-navigation/issues/298 however that issue applies to StackNavigation, and I can understand how in a stack navigation scenario the component might never unmount. However I would assume in a drawer navigation it would?
If this is by design my mistake, I guess the option is to create a custom navigator that returns different components. In which case they should definitely unmount I assume?
Thanks

Edit: So when I navigate away from the component I get no componentWillUnmount, however when I navigate back to it I get componentWillMount. This is confusing as how am I getting componentWillMount but no componentWillUnmount?

**Comment by dantman** [comment]
I just tried this in my own app and unmount seems to be working fine.
I added a `{routeName} mounted` log to mount and a `{routeName} unmounted` log to unmount.

Switching through drawer items I got:
* A mounted
* *Select B*
* B mounted
* A unmounted
* *Select C*
* C mounted
* B unmounted
* *Select A*
* A mounted
* C unmounted

**Comment by SamMatthewsIsACommonName** [comment]
I'm sorry I just realised I should have framed this issue differently, my drawer navigation returns several stack navigations, however the first component of each stack navigation is never being unmounted when I switch between them. Does that make sense?

**Comment by yomybaby** [comment]
I have a same issue at the same case.
componentWillUnmount never happens with StackNavigation nested inside DrawerNavigation.
Any help?

**Comment by aaronbuchanan** [comment]
Same issue, halp!

**Comment by galhardi** [comment]
+1

**Comment by csumrell** [comment]
I also have this issue

**Comment by danaszova** [comment]
+1

**Comment by timLoewel** [comment]
I have a TabNavigator nested inside a StackNavigator.
When something is pushed on top of the TabNavigator, the component currently visible in the TabNavigator does not get a componentWillUnMount event. But it gets a componentWillMount event, as soon as it gets back into view.

**Comment by locales-wellington** [comment]
+1

**Comment by danaszova** [comment]
I have this issue as well.

**Comment by kelset** [comment]
Can anyone provide a small code sample (using [Snack](snack.expo.io) maybe) that reproduce the issue?

**Comment by kelset** [comment]
Pinging OP @SamMatthewsIsACommonName since this issue is quite old and related to an old version of the lib. If no answer I'll close in 7 days.

**Comment by patlux** [comment]
@SamMatthewsIsACommonName 

I think it's by design. Your components remain mounted as long as they are in the history tree. To avoid re-renders and recalculations of your screens while they are not visible by the user, use libraries like redux and reselect.

**Comment by SamMatthewsIsACommonName** [comment]
@patlux I understand how its helpful for performance but it really strips a react component of one of its foundational and useful lifecycle hooks. For example a details screen that uses a dynamic listener which should be cleaned up when the component unmounts. Instead need to use componentWillMount or componentDidMount to remove the previous listener if it exists which I just find a bit weird (this is the strange crux of the issue, that both of these would be called however componentWillUnmount is never called). Unfortunately it's incredibly difficult to recreate a code example as our app is huge with around 60 screens... it would take me days to do this sorry.

**Comment by patlux** [comment]
If you want to use the lifecycle hooks, then create a hoc which checks if  the screen is currently visible or not. If not, then return null. After returning null the wrapped component will be unmounted.

Small example:
```javascript
class HomeContainer extends React.Component {

  componentWillUnmount() {
    console.log('Screen exit');
  }

}

class HomeScreen extends React.Component {

  render() {
    if (/* check here if the screen is not visible */) {
      return null;
    }

    return <HomeContainer />;
  }

}

export default HomeScreen;
```

**Comment by SamMatthewsIsACommonName** [comment]
I see thanks, so would you check if the component is visible using the redux state of the navigation and comparing that to the id of the route or something?

**Comment by patlux** [comment]
Yes, correctly. You can check for more examples this: [react-navigation-is-focused-hoc](https://github.com/patlux/react-navigation-is-focused-hoc) It's not for the use with Redux. But the idea is the same.

**Comment by andreasanta** [comment]
Hi everyone, stumbled upon this thread while trying to intercept focus event for a screen.

Pardon the ignorant question, but I need to handle the "focused" event in three cases:

-> User presses back button on titlebar
-> User presses Android hardware back button
-> User presses my custom button to go back

Is this currently possible? Will this FR address this case?

**Comment by locales-wellington** [comment]
@patlux it also seems componentWillMount / didMount is called every time screen is displayed? I have a situation where I load a map on mount and it loads a new one every time the screen is displayed. This seems like inconsistent behaviour, if the component remains mounted, shouldn't this only fire once?

**Comment by SamMatthewsIsACommonName** [comment]
This is the part that really puzzles me... willMount / didMount but no unMount ?? It's just strange

**Comment by 0kai** [comment]
I add an emitter to notify the route changed.
``` javascript
// navigator
<YourNavigator onNavigationStateChange={(prevState, currentState) => {
    let route = currentState;
    while (route.routes) {
      route = route.routes[route.index];
    }
    DeviceEventEmitter.emit('routeStateChanged', route);
  }}
  />

// component
  componentDidMount() {
    this.routeSubscription = DeviceEventEmitter.addListener('routeStateChanged', this.onRouteStateChanged);
  }

  componentWillUnmount() {
    this.routeSubscription.remove();
  }

  onRouteStateChanged = (route) => {
    const { state } = this.props.navigation;
    if (state.routeName === route.routeName && state.key === route.key) {
      // resume
    } else {
      // pause
    }
  }
```

**Comment by patlux** [comment]
@locales-wellington: Can you please provide a [snack](https://snack.expo.io/) which shows this behaviour? That's not normal. Never happens to me. I know it will take some time to create a snack, but if you want it to be fixed...

**Comment by ArvoGuo** [comment]
+1

**Comment by locales-wellington** [comment]
@patlux I've done a snack https://snack.expo.io/S1xM7hJAW Seems to be a specific edge case where you have an inline navigation from drawer to a stack the unmount is not called. Hope that helps.

**Comment by rochapablo** [comment]
I'm not sure, but I'm facing a problem that could be related;

I'm using ```DeviceEventEmitter``` to emit some events and find out that they are been fired twice.

I thought that I could be doing something wrong (maybe I still) but when I set ```initialRouteName``` to render my screen (as default) besides redirect ```(this.props.navigation.navigate)``` using some button, the event it's fired only once.

**Comment by rochapablo** [comment]
@kelset, [here's an example](https://bitbucket.org/pablo_rocha/backpack)

At the first access will work:

```
[ 'Backpack', 'Login.componentDidMount' ]
[ 'Backpack', 'Base.componentDidMount' ]
[ 'Backpack', 'Dashboard.componentDidMount' ]
[ 'Backpack', 'Cricket.componentDidMount' ]
[ 'Backpack', 'Sidenav.componentDidMount' ]
[ 'Backpack', 'Base.messageListener', { name: 'Dashboard!' } ]
[ 'Backpack', 'Cricket.listener', { name: 'Dashboard!' } ]
```

After that, clicking logout, will start duplicate:

```
[ 'Backpack', 'Login.componentDidMount' ]
[ 'Backpack', 'Login.componentDidMount' ]
[ 'Backpack', 'Base.componentDidMount' ]
[ 'Backpack', 'Dashboard.componentDidMount' ]
[ 'Backpack', 'Cricket.componentDidMount' ]
[ 'Backpack', 'Sidenav.componentDidMount' ]
[ 'Backpack', 'Base.messageListener', { name: 'Dashboard!' } ]
[ 'Backpack', 'Base.messageListener', { name: 'Dashboard!' } ]
[ 'Backpack', 'Cricket.listener', { name: 'Dashboard!' } ]
[ 'Backpack', 'Cricket.listener', { name: 'Dashboard!' } ]
```

**Comment by kelset** [comment]
Thanks for keeping us updated @rochapablo , much appreciated (btw it may be linked to this issue https://github.com/react-community/react-navigation/issues/2599 - writing this comment just to "wire" them)

**Comment by rochapablo** [comment]
Just keeping update:

[Following](https://github.com/react-community/react-navigation/issues/2599#issuecomment-337845786) @Raveler didn't work.

If I want go back to another child route, the problem will persist.

Login/Emit Message/Register/Submit/Emit Message

```
[ 'Backpack', 'Base.componentDidMount' ]
[ 'Backpack', 'Dashboard.componentDidMount' ]
[ 'Backpack', 'Cricket.componentDidMount' ]
[ 'Backpack', 'Sidenav.componentDidMount' ]
[ 'Backpack', 'Base.messageListener', { name: 'Dashboard!' } ]  // OKAY
[ 'Backpack', 'Cricket.listener', { name: 'Dashboard!' } ]  // OKAY

[ 'Backpack', 'Login.componentDidMount' ]
[ 'Backpack', 'Register.componentDidMount' ]
[ 'Backpack', 'Base.componentDidMount' ]
[ 'Backpack', 'Dashboard.componentDidMount' ]
[ 'Backpack', 'Cricket.componentDidMount' ]
[ 'Backpack', 'Sidenav.componentDidMount' ]
[ 'Backpack', 'Base.messageListener', { name: 'Dashboard!' } ]  // OKAY
[ 'Backpack', 'Base.messageListener', { name: 'Dashboard!' } ] // NOT OKAY
[ 'Backpack', 'Cricket.listener', { name: 'Dashboard!' } ]  // OKAY
[ 'Backpack', 'Cricket.listener', { name: 'Dashboard!' } ] // NOT OKAY
```

I've also update [the example](https://bitbucket.org/pablo_rocha/backpack) if needed.

**Comment by dbyilmaz** [comment]
@SamMatthewsIsACommonName  did you solve this problem?

**Comment by embpdaniel** [comment]
I think I am having the same issue. I have a Drawer component (from 'react-nativedrawer') which renders a StackNavigator:

```javascript
const AppNavigator = StackNavigator(
  {
    Bookmarks: { screen: Bookmarks },
    CorporateDirectory: { screen: CorporateDirectory }
  },
  {
    headerMode: 'none'
  }
);
...
render() {
    const { setMenuOpened } = this.props;
    return (
      <Drawer
        ref={ref => (this._drawer = ref)}
        type={'displace'}
        open={this.props.app.menuOpened}
        acceptTap={true}
        tapToClose={true}
        panOpenMask={0.1}
        panCloseMask={0.2}
        acceptPan={true}
        panThreshold={0.25}
        content={this._menuContent()}
        openDrawerOffset={150}
        tweenHandlerOn={false}
        tweenDuration={250}
        tweenEasing={'linear'}
        closedDrawerOffset={-3}
        relativeDrag={false}
        negotiatePan={false}
        useInteractionManager={true}
        styles={drawerStyles}
        onOpen={() => setMenuOpened(true)}
        onClose={() => setMenuOpened(false)}
      >
        <AppNavigator ref={nav => (this.navigator = nav)} />
      </Drawer>
    );
  }
```
When Bookmarks opens, component mounts. When I navigate to CorporateDirectory, component never un-mounts. When I navigate back to Bookmarks, component mounts again.

**Comment by heroofchina** [comment]
Me too,but I think the reason is that  react-navigation need  to track the routerstack ,so the component is save in the route stack but no  be destroyed;

**Comment by brentvatne** [comment]
can someone try to phrase this problem succinctly and provide an example? much appreciated!

**Comment by sebastienbarbier** [comment]
This seams to be only when using `.navigate('myState')`. Replacing with `goBack()` or `.popToTop('myState')` works for me.

**Comment by brentvatne** [comment]
@sebastienbarbier - you can configure SwitchNavigator with `resetOnBlur` and `backBehavior` options:

```
let Switch = SwitchNavigator(routes, { resetOnBlur: false, backBehavior: 'initialRoute' })
```

this will make it behave like a tab navigator. however to make back work chronologically, rather than always going back to the first route, we need to implement this rfc https://github.com/react-navigation/rfcs/pull/2

**Comment by brentvatne** [comment]
I'm closing this issue as it doesn't clearly describe a bug. Please create a new issue if you think this is a bug and I am missing something, and follow the template so I can understand it, thanks!

**Comment by brentvatne** [comment]
also, the original issue as per the title is resolved in 2.0 since we implemented this rfc: https://github.com/react-navigation/rfcs/blob/master/text/0004-less-pushy-navigate.md

2.0 rc is available right now on npm (latest is 2.0.0-rc.5 at the time of writing). full release expected this week
